problems/problems.go

```diff
@@ -2,73 +2,146 @@ package errors
+
+//Problem implements the RFC 7807 "problem"/error standard
 //Additional error field can be defined in custom struct implementing the Problem struct
 type Problem struct {
 	Title    string `json:"title"`
 	Detail   string `json:"detail,omitempty"`
 	Type     string `json:"type,omitempty"`
 	Instance string `json:"instance,omitempty"`
 	Status   int    `json:"status,omitempty"`
+	cause    error
 }
 
-//Error implements the error interface
-func (p *Problem) Error() string {
-	return fmt.Sprintf("Title: %s; Detail: %s", p.Title, p.Detail)
+//Wrap is an alias to github.com/pkg/errors Wrap function. If Problem pointer passed as error, it sets the error and a new message of Problem.
+func Wrap(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+
+	problem, ok := err.(*Problem)
+	if !ok {
+		return errors.Wrap(err, msg)
+	}
+
+	problem.cause = errors.Wrap(problem.cause, msg)
+	return problem
 }
 
-//DefaultType implements the default type content described in RFC7807
-const DefaultType = "about:blank"
+//WithStack is an alias to github.com/pkg/errors WithStack function. If Problem pointer passed as error, it sets the error of Problem.
+func WithStack(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	problem, ok := err.(*Problem)
+	if !ok {
+		return errors.WithStack(err)
+	}
+
+	problem.cause = errors.WithStack(problem.cause)
+	return problem
+}
+
+//WithMessage is an alias to github.com/pkg/errors WithMessage function. If Problem pointer passed as error, it sets the error, stacktrace and message of Problem.
+func WithMessage(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+
+	problem, ok := err.(*Problem)
+	if !ok {
+		return errors.WithMessage(err, msg)
+	}
+
+	problem.cause = errors.WithMessage(problem.cause, msg)
+	return problem
+}
+
+//New creates a new Problem with all object details described in RFC 7807
+func New(title, detail string, status int) error {
+	return newWithError(
+		errors.New(detail),
+		title,
+		detail,
+		status,
+	)
+}
 
-//SetTitle sets the title of the problem object
-func (p *Problem) SetTitle(title string) *Problem {
+func newWithError(err error, title, detail string, status int) *Problem {
+	return &Problem{
+		Type:   DefaultType,
+		Title:  title,
+		Detail: err.Error(),
+		Status: status,
+		cause:  err,
+	}
+}
+
+//Format displays the contained error in a specific format
+func (p Problem) Format(s fmt.State, verb rune) {
+	if p.cause == nil {
+		fmt.Fprint(s, p.Detail)
+		return
+	}
+
+	formatter, ok := p.cause.(fmt.Formatter)
+	if !ok {
+		fmt.Fprint(s, p.Detail)
+		return
+	}
+
+	formatter.Format(s, verb)
+}
+
+//Cause returns the outermost error of Problem
+func (p Problem) Cause() error {
+	return p.cause
+}
+
+//Error returns the error of the problem
+func (p Problem) Error() string {
+	return p.cause.Error()
+}
+
+//SetTitle sets the title field of the problem, specified in RFC 7807
+func (p Problem) SetTitle(title string) *Problem {
```

Comment:
why return a pointer for all those helpers if they themselves only are non-pointer access?